role/chore: feat to run fluentd as another group

StessyMougeot opened the pull request.

Commit:
role/chore: feat to run fluentd as another group, used to parse var/log, syslog, …

ghost commented:
Un petit update dans le README.md serait le bienvenu, ainsi qu'un fichier changelog et un tag, comme pour les autres rôles ;)

sterchelen commented on tasks/main.yml:
```diff
@@ -10,6 +10,14 @@
     name: td-agent
     state: "{{ fluentd_package_state }}"
 
+- name: Change group to adm
```
```suggestion
- name: Override fluentd default group to {{ fluentd_group }}
```

Commit:
role/chore: update readme with new var

sterchelen commented on tasks/main.yml:
```diff
@@ -10,6 +10,14 @@
     state: "{{ fluentd_package_state }}"
 
+- name: Change group to adm
+  lineinfile:
+    path: /etc/systemd/system/multi-user.target.wants/td-agent.service
+    regexp: '^Group='
+    line: 'Group={ fluentd_group }'
```
If you look closely to this unit file
```
[Unit]
Description=td-agent: Fluentd based data collector for Treasure Data
Documentation=https://docs.treasuredata.com/articles/td-agent
After=network-online.target
Wants=network-online.target

[Service]
User=td-agent
Group=adm
LimitNOFILE=65536
Environment=LD_PRELOAD=/opt/td-agent/lib/libjemalloc.so
Environment=GEM_HOME=/opt/td-agent/lib/ruby/gems/2.7.0/
Environment=GEM_PATH=/opt/td-agent/lib/ruby/gems/2.7.0/
Environment=FLUENT_CONF=/etc/td-agent/td-agent.conf
Environment=FLUENT_PLUGIN=/etc/td-agent/plugin
Environment=FLUENT_SOCKET=/var/run/td-agent/td-agent.sock
Environment=TD_AGENT_LOG_FILE=/var/log/td-agent/td-agent.log
Environment=TD_AGENT_OPTIONS=
EnvironmentFile=-/etc/default/td-agent
PIDFile=/var/run/td-agent/td-agent.pid
RuntimeDirectory=td-agent
Type=forking
# XXX: Fix fluentd executables path
ExecStart=/opt/td-agent/bin/fluentd --log $TD_AGENT_LOG_FILE --daemon /var/run/td-agent/td-agent.pid $TD_AGENT_OPTIONS
ExecStop=/bin/kill -TERM ${MAINPID}
ExecReload=/bin/kill -HUP ${MAINPID}
Restart=always
TimeoutStopSec=120

[Install]
WantedBy=multi-user.target
```

you have this mention `EnvironmentFile=-/etc/default/td-agent`
Let's add the group into the `/etc/default/td-agent` file. Would be much more maintainable :)

StessyMougeot replied:
Not possible, it's an escalation privileges

sterchelen replied:
Seen with you privately. Let's add `td-agent` user to the group `adm`. This way, we don't have to modify the service file :smiley:

Commit:
role/chore: Ahh changelog

sterchelen commented on CHANGELOG.md:
```diff
@@ -0,0 +1,7 @@
+Changelog
+=========
+
```
Could you add `0.1.1` with the actual changes?

Commit:
chore/changelog: 0.1.1

Commit:
role/groups: Allow multiple groups, and allow vault env var
-- empty line -- (github needs an empty line to collapse the next section on the UI)

Define multiple groups to add
Possibility to override unit file to add vault env usage

Commit:
role/chore: update changelog

sterchelen commented on README.md:
```diff
@@ -14,12 +14,22 @@ Available variables are listed below, along with default values (see `defaults/m
 
     fluentd_package_state: present
 
+Extra groups to add log access, secret:
```
What you mean by `, secret`?

StessyMougeot replied:
For exemple, to read datadog config file for vault, user must be in `vault-datadog` group

sterchelen replied:
Then I would rephrase like this:
```suggestion
Extra groups to give specific accesses to fluentd user, like logs or secret:
```

sterchelen commented on README.md:
```diff
@@ -14,12 +14,22 @@ Available variables are listed below, along with default values (see `defaults/m
+Vault integration to use alongside `aggregator` tag to load secret in environment variable:
+    fluentd_vault_config_path:
+    fluentd_secret_path:
+    fluent_secret_name:
```
```suggestion
    fluentd_secret_name:
```

sterchelen commented on tasks/main.yml:
```diff
@@ -10,6 +10,15 @@
+- name: "Add user to {{ fluentd_groups | join(',') }} group"
+  user:
+    name: td-agent
+    group: td-agent
```
Not sure if it is mandatory?

sterchelen replied:
I mean specifying `group: td-agent`

StessyMougeot replied:
I don't think it's mandatory, but for informations about primary group

sterchelen commented on tasks/main.yml:
```diff
@@ -28,8 +37,54 @@
   with_items: "{{ fluentd_plugins }}"
 
+- name: retrieve initial ExecStart
+  shell: 'grep -i "ExecStart=.*/.*" /lib/systemd/system/multi-user.target.wants/td-agent.service | sed -r "s/^([^\/])*(\/.*)/\2/"'
```
Is it possible to use `/lib/systemd/system/td-agent.service`? We should use the original service file

StessyMougeot replied:
Sure, copy/paste mistake

sterchelen commented on tasks/main.yml:
```diff
@@ -28,8 +37,54 @@
+  - never
+  - aggregator
+
+- name: Ensure override file is absent
+  file:
+    path: /etc/systemd/system/td-agent.service.d/override.conf
+    state: absent
+  tags:
+  - never
+  - aggregator
```
Is it mandatory to ensure if the file is absent?

StessyMougeot replied:
Is there a better way to clear the file with ansible?

sterchelen replied:
Yes, using a template file :)

sterchelen replied:
Then you can remove merge `reset ExecStart` task with `Add vault secret`

StessyMougeot replied:
Much better, thanks

sterchelen commented on tasks/main.yml:
```diff
@@ -28,8 +37,54 @@
+    path: /etc/systemd/system/td-agent.service.d/override.conf
+    state: present
+    insertafter: '^ExecStart=$'
+    line: 'ExecStart=vault --config-file {{ fluentd_vault_config_path }} env -p {{ fluentd_secret_path }}{{ fluent_secret_name }} -- {{ initial_cmd.stdout }}'
```
To improve readability:
```suggestion
    line: > 
    'ExecStart=vault --config-file {{ fluentd_vault_config_path }}
    env -p {{ fluentd_secret_path }}{{ fluent_secret_name }} -- {{ initial_cmd.stdout }}'
```

Commit:
Update README.md

Co-authored-by: Nicolas Sterchele <nicolas_sterchele@ultimatesoftware.com>

Commit:
Improve reliability

Co-authored-by: Nicolas Sterchele <nicolas_sterchele@ultimatesoftware.com>

sterchelen commented on README.md:
```diff
@@ -14,12 +14,22 @@ Available variables are listed below, along with default values (see `defaults/m
     fluentd_package_state: present
 
+Extra groups to add log access, secret:
+    fluentd_groups:
```
Your naming `extra groups` speaks by itself:
```suggestion
    fluentd_extra_groups:
```

StessyMougeot replied:
I've hesitated between both in fact

sterchelen commented on README.md:
```diff
@@ -14,12 +14,22 @@ Available variables are listed below, along with default values (see `defaults/m
+
 The `td-agent` Fluentd package state; set to `latest` to upgrade or change versions.
 
+    fluentd_adm_group: false
```
I am not seeing where you are using it :eyes:

StessyMougeot replied:
Not anymore, incorporated in `fluentd_extra_groups`

Commit:
role/fix: wrong path to look for system unit

Commit:
Merge branch 'feat-group-system' of github.com:peopledoc/ansible-role-fluentd into feat-group-system

Commit:
role/fix: typo and unused var

Commit:
role/chore: improve Readme explanation

Co-authored-by: Nicolas Sterchele <nicolas_sterchele@ultimatesoftware.com>

Commit:
role/default: update default vars

Commit:
Merge branch 'feat-group-system' of github.com:peopledoc/ansible-role-fluentd into feat-group-system

Commit:
role/refactor: use template file

sterchelen commented on tasks/main.yml:
```diff
@@ -28,8 +37,33 @@
+
+- name: Add unit override.conf file
+  template:
+    src: templates/override.conf.j2
```
You've forgotten to add the template file :smiley:

Commit:
role/refactor: add template
